File: sim/src/grid.ts
export type HtmlBox = {
  left: number;
  top: number;
  width: number;
  height: number;
};

export type BarLayout = {
  columns: number;
  rows: number;
  width: number;
  height: number;
};

export type Handle = "n" | "s" | "e" | "w" | "ne" | "nw" | "se" | "sw";

export const HANDLES: Handle[] = ["n", "s", "e", "w", "ne", "nw", "se", "sw"];

export function columnChoices(buttons: number): number[] {
  const count = Math.max(1, buttons);
  const cols: number[] = [];
  for (let c = count; c >= 1; c--) {
    if (count % c === 0) cols.push(c);
  }
  return cols;
}

export function layoutForColumns(buttons: number, columns: number, slotSize: number): BarLayout {
  const count = Math.max(1, buttons);
  const cols = Math.max(1, Math.min(columns, count));
  const rows = Math.ceil(count / cols);
  return {
    columns: cols,
    rows,
    width: cols * slotSize,
    height: rows * slotSize,
  };
}

export function nearestBarLayout(
  buttons: number,
  slotSize: number,
  width: number,
  height: number,
): BarLayout {
  const aspect = Math.max(width, 1) / Math.max(height, 1);
  const logAspect = Math.log(aspect);
  let best = layoutForColumns(buttons, columnChoices(buttons)[0], slotSize);
  let bestDist = Number.POSITIVE_INFINITY;
  columnChoices(buttons).forEach((cols) => {
    const layout = layoutForColumns(buttons, cols, slotSize);
    const layoutLog = Math.log(layout.width / layout.height);
    const dist = (layoutLog - logAspect) * (layoutLog - logAspect);
    if (dist < bestDist) {
      bestDist = dist;
      best = layout;
    }
  });
  return best;
}

export function snapValue(value: number, grid: number): number {
  return Math.round(value / grid) * grid;
}

export function snapBox(box: HtmlBox, grid: number): HtmlBox {
  return {
    left: snapValue(box.left, grid),
    top: snapValue(box.top, grid),
    width: Math.max(grid, snapValue(box.width, grid)),
    height: Math.max(grid, snapValue(box.height, grid)),
  };
}

export function snapChromeSize(width: number, height: number, grid: number): { width: number; height: number } {
  return {
    width: Math.max(grid, snapValue(width, grid)),
    height: Math.max(grid, snapValue(height, grid)),
  };
}

export function applyHandleSize(box: HtmlBox, handle: Handle, width: number, height: number): HtmlBox {
  let left = box.left;
  let top = box.top;
  if (handle.indexOf("w") !== -1) left = box.left + box.width - width;
  if (handle.indexOf("n") !== -1) top = box.top + box.height - height;
  return { left, top, width, height };
}

export function desiredSize(start: HtmlBox, handle: Handle, dx: number, dy: number): { width: number; height: number } {
  let width = start.width;
  let height = start.height;
  if (handle.indexOf("e") !== -1) width = start.width + dx;
  if (handle.indexOf("w") !== -1) width = start.width - dx;
  if (handle.indexOf("s") !== -1) height = start.height + dy;
  if (handle.indexOf("n") !== -1) height = start.height - dy;
  return { width, height };
}

File: bars/grid.lua
--[[
  Purpose: Snap a Bar drag size to a columns/grid that fills the slot count.
  Deps: ShadowUI addon table
  Public: ColumnChoices(), SnapBarColumns(), BarGridMetrics(), LayoutForColumns(),
          NearestBarLayout()
]]

local Addon = LibStub("AceAddon-3.0"):GetAddon("ShadowUI")

function Addon:ColumnChoices(buttons)
  local count = math.max(1, buttons or 1)
  local cols = {}
  for c = count, 1, -1 do
    if count % c == 0 then
      cols[#cols + 1] = c
    end
  end
  return cols
end

function Addon:SnapBarColumns(buttons, columns)
  local choices = self:ColumnChoices(buttons)
  local want = columns or choices[1]
  local best = choices[1]
  local bestDist = math.abs(best - want)
  for _, cols in ipairs(choices) do
    local dist = math.abs(cols - want)
    if dist < bestDist then
      best = cols
      bestDist = dist
    end
  end
  return best
end

function Addon:BarGridMetrics(buttons, columns, slotSize, gap, rowGaps)
  local count = math.max(1, buttons or 1)
  local cols = math.max(1, math.min(columns or count, count))
  local rows = math.ceil(count / cols)
  local size = slotSize or 36
  local space = gap or 0
  local rowTop = {}
  local rowPad = {}
  for r = 1, rows do
    rowTop[r] = (r - 1) * (size + space)
    local rg = rowGaps and rowGaps[r] or {}
    local above = math.max(0, rg.above or 0)
    local below = math.max(0, rg.below or 0)
    local pad = above + below
    local icon = size
    if pad >= size then
      local scale = (size - 1) / pad
      above = above * scale
      below = below * scale
      icon = 1
    elseif pad > 0 then
      icon = size - pad
    end
    rowPad[r] = { above = above, icon = icon }
  end
  return {
    columns = cols,
    rows = rows,
    width = cols * size + (cols - 1) * space,
    height = rows * size + (rows - 1) * space,
    rowTop = rowTop,
    rowPad = rowPad,
  }
end

function Addon:LayoutForColumns(buttons, columns, slotSize, gap, rowGaps)
  local metrics = self:BarGridMetrics(buttons, columns, slotSize, gap, rowGaps)
  return {
    columns = metrics.columns,
    rows = metrics.rows,
    width = metrics.width,
    height = metrics.height,
  }
end

function Addon:NearestBarLayout(buttons, slotSize, width, height, gap, rowGaps)
  local aspect = math.max(width or 1, 1) / math.max(height or 1, 1)
  local logAspect = math.log(aspect)
  local choices = self:ColumnChoices(buttons)
  local best = self:LayoutForColumns(buttons, choices[1], slotSize, gap, rowGaps)
  local bestDist = math.huge
  for _, cols in ipairs(choices) do
    local layout = self:LayoutForColumns(buttons, cols, slotSize, gap, rowGaps)
    local layoutLog = math.log(layout.width / layout.height)
    local dist = (layoutLog - logAspect) * (layoutLog - logAspect)
    if dist < bestDist then
      bestDist = dist
      best = layout
    end
  end
  return best
end
